scipio/src/file_stream.rs

```diff
@@ -615,12 +615,355 @@ impl Future for PrepareBuffer {
+/// as a position must always be specified.
+///
+/// Very rarely does one need to issue random writes to a file. Therefore, the [`StreamWriter`] is
+/// likely your go-to API when it comes to writing files.
+///
+/// The [`StreamWriter`] implements [`AsyncWrite`]. Because it is backed by a Direct I/O
+/// file, the flush method has no effect. Closing the file issues a sync so that
+/// the data can be flushed from the internal NVMe caches.
+///
+/// [`DmaFile`]: struct.DmaFile.html
+/// [`StreamReader`]: struct.StreamReader.html
+/// [`get_buffer_aligned`]: struct.StreamReader.html#method.get_buffer_aligned
+/// [`AsyncWrite`]: https://docs.rs/futures/0.3.5/futures/io/trait.AsyncWrite.html
+pub struct StreamWriter {
+    file: Option<Rc<DmaFile>>,
+    state: Rc<RefCell<StreamWriterState>>,
+}
+
+impl StreamWriter {
+    fn new(builder: StreamWriterBuilder) -> StreamWriter {
+        let mut buffer_queue = VecDeque::with_capacity(builder.write_behind);
+        for _ in 0..builder.write_behind {
+            buffer_queue.push_back(Rc::new(DmaFile::alloc_dma_buffer(builder.buffer_size)));
+        }
+
+        let state = StreamWriterState {
+            buffer_size: builder.buffer_size,
+            write_behind: builder.write_behind,
+            flush_on_close: builder.flush_on_close,
+            buffer_queue,
+            waker: None,
+            pending: Vec::new(),
+            error: None,
+            file_status: FileStatus::Open,
+            file_pos: 0,
+            buffer_pos: 0,
+        };
+
+        let state = Rc::new(RefCell::new(state));
+        StreamWriter {
+            file: Some(builder.file),
+            state,
+        }
+    }
+
+    /// Acquires the current position of this [`StreamWriter`].
+    ///
+    /// # Examples
+    /// ```no_run
+    /// use scipio::{DmaFile, StreamWriterBuilder, LocalExecutor};
+    /// use futures::io::AsyncWriteExt;
+    ///
+    /// let ex = LocalExecutor::make_default();
+    /// ex.run(async {
+    ///     let file = DmaFile::create("myfile.txt").await.unwrap();
+    ///     let mut writer = StreamWriterBuilder::new(file).build();
+    ///     assert_eq!(writer.current_pos(), 0);
+    ///     writer.write_all(&[0, 1, 2, 3, 4]).await.unwrap();
+    ///     assert_eq!(writer.current_pos(), 5);
+    ///     writer.close().await.unwrap();
+    /// });
+    /// ```
+    ///
+    /// [`StreamWriter`]: struct.StreamWriter.html
+    pub fn current_pos(&self) -> u64 {
+        self.state.borrow().file_pos()
+    }
+}
+
+impl AsyncWrite for StreamWriter {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        let mut state = self.state.borrow_mut();
+        if let Some(err) = current_error!(state) {
+            return Poll::Ready(err);
+        }
+
+        let mut written = 0;
+        while written < buf.len() {
+            match state.buffer_queue.pop_front() {
```

Comment:
If I understand correctly, if my `StreamWriter` was configured with `n` write-behind buffers and I try to issue `n+1` write operations (assuming my writes are of the same buffer size), then the CPU will spin until a buffer is available (i.e until one of my previous writes has completed)?

Assuming this is correct, why not use our own async dequeue here?

Reply:
That's note quite right.

First the whole point of this API is that you don't try to issue any operations. You just write bytes linearly to the file without worrying much about what the storage is doing. Writes are done to the storage as necessary.

But if you have already `n` in-flight buffers and need to write another, it doesn't spin. It just returns `Poll::Pending`, which tells the rust core to go do something else.

The AsyncDeque is a good way to implement this in contexts outside `poll`. But inside `poll` you just return a status saying whether or not you are ready.

Reply:
Alright, thanks for the explanation :bow: